src/refiners/fluxion/layers/attentions.py

```diff
@@ -18,68 +18,137 @@ def scaled_dot_product_attention(
+    """
+    return _scaled_dot_product_attention(
+        query=query,
+        key=key,
+        value=value,
+        is_causal=is_causal,
+    )
 
-def sparse_dot_product_attention_non_optimized(
+
+def scaled_dot_product_attention_non_optimized(
     query: Float[Tensor, "batch source_sequence_length dim"],
     key: Float[Tensor, "batch target_sequence_length dim"],
     value: Float[Tensor, "batch target_sequence_length dim"],
     is_causal: bool = False,
 ) -> Float[Tensor, "batch source_sequence_length dim"]:
+    """Non-optimized Scaled Dot Product Attention.
+
+    See [[arXiv:1706.03762] Attention Is All You Need (Equation 1)](https://arxiv.org/abs/1706.03762) for more details.
+    """
     if is_causal:
         # TODO: implement causal attention
-        raise NotImplementedError("Causal attention for non_optimized attention is not yet implemented")
-    _, _, _, dim = query.shape
+        raise NotImplementedError(
+            "Causal attention for `scaled_dot_product_attention_non_optimized` is not yet implemented"
+        )
+
+    dim = query.shape[-1]
     attention = query @ key.permute(0, 1, 3, 2)
     attention = attention / math.sqrt(dim)
     attention = torch.softmax(input=attention, dim=-1)
     return attention @ value
 
 
 class ScaledDotProductAttention(Module):
+    """Scaled Dot Product Attention.
+
+    ??? note "See [[arXiv:1706.03762] Attention Is All You Need (Figure 2)](https://arxiv.org/abs/1706.03762) for more details"
+
+        ![](https://ar5iv.labs.arxiv.org/html/1706.03762/assets/Figures/ModalNet-19.png)
+
+    Note:
+        This layer simply wraps `scaled_dot_product_attention` inside an `fl.Module`.
+
+    Receives:
+        Query (Float[Tensor, "batch num_queries embedding_dim"]):
+        Key (Float[Tensor, "batch num_keys embedding_dim"]):
+        Value (Float[Tensor, "batch num_values embedding_dim"]):
+
+    Returns:
+        (Float[Tensor, "batch num_queries embedding_dim"]):
+
+    Example:
+        ```py
+        attention = fl.ScaledDotProductAttention(num_heads=8)
+
+        query = torch.randn(2, 10, 128)
+        key = torch.randn(2, 10, 128)
+        value = torch.randn(2, 10, 128)
+        output = attention(query, key, value)
+
+        assert output.shape == (2, 10, 128)
+        ```
+    """
+
     def __init__(
         self,
         num_heads: int = 1,
-        is_causal: bool | None = None,
+        is_causal: bool = False,
         is_optimized: bool = True,
         slice_size: int | None = None,
     ) -> None:
+        """Initialize the Scaled Dot Product Attention layer.
+
+        Args:
+            num_heads: The number of heads to use.
+            is_causal: Whether to use causal attention.
+            is_optimized: Whether to use optimized attention.
+            slice_size: The slice size to use for the optimized attention.
```

Comment:
E.g. here too, re: defaults, it would at least be great to state that this is optional.